### PR TITLE
bugfix:lookupcache's Get method can not be called concurrently

### PR DESCRIPTION
**Pull request by mqliang**
ref https://github.com/kubernetes/kubernetes/issues/26376

@lavalamp @therc @mikedanese

### CODE REVIEW

**Comment by mqliang** [comment]
cc @249043822

**Comment by therc** [comment]
I've rebuilt my binaries with this change. No panics so far. Previously, it would happen as early as ten minutes after starting.

**Comment by lavalamp** [comment]
LGTM, thanks!

File: pkg/controller/lookup_cache.go [hunk 1]
```diff
@@ -72,8 +72,9 @@ func (c *MatchingCache) Add(labelObj objectWithMeta, selectorObj objectWithMeta)
 // we need check in the external request to ensure the cache data is not dirty.
 func (c *MatchingCache) GetMatchingObject(labelObj objectWithMeta) (controller interface{}, exists bool) {
 	key := keyFunc(labelObj)
-	c.mutex.RLock()
-	defer c.mutex.RUnlock()
+	// NOTE: we use Lock() instead of RLock() here because lru's Get() method is not threadsafe
```
Comment (xiang90) [inline pkg/controller/lookup_cache.go]: Get() method modifies state? If it does not modify the state, RLock is fine. We do not need it to be threadsafe. That is the whole point of the lock here.

**Comment by xiang90** [comment]
@lavalamp I think we need to fix the comment. it is confusing i think.

File: pkg/controller/lookup_cache.go (hunk 1 above)
Comment (mqliang) [inline pkg/controller/lookup_cache.go]: SGTM,  I will make the comment more specific

**Comment by dims** [comment]
@xiang90 : if you see Get method in lru
https://github.com/golang/groupcache/blob/master/lru/lru.go#L74

you will see that it calls MoveToFront in
https://golang.org/src/container/list/list.go

which modifies the linked list. So we can't use just RLock

File: pkg/controller/lookup_cache.go (hunk 1 above)
Comment (lavalamp) [inline pkg/controller/lookup_cache.go]: It is an LRU cache, so yes, Get() modifies state-- LRU caches move the most recently accessed item to the head of a linked list.

**Comment by xiang90** [comment]
@dims @mqliang Actually I understand it modifies the state. I just want to say we should modify the comment to `Get() method modifies state` instead of `thread-safe` :) Thanks!

**Comment by dims** [comment]
@xiang90 Aha! +1000 if we can go fix it in a reasonable time frame. For now, we just have to live with it.

**Comment by mqliang** [comment]
@lavalamp I have updated this PR and make the comment more specific as @xiang90 suggested. Could you please reapply the LGTM label?

**Comment by xiang90** [comment]
LGTM. Thanks @mqliang !

**Pull request merged**